File: tests/typesafety/python/higher_order.py
# expect: passes
# title: After retry, fetch keeps its full signature
# improves: retry_any
from collections.abc import Callable


def retry[**P, R](times: int, operation: Callable[P, R]) -> Callable[P, R]:
    def attempt(*args: P.args, **kwargs: P.kwargs) -> R:
        failures = 0
        while True:
            try:
                return operation(*args, **kwargs)
            except ConnectionError:
                failures += 1
                if failures >= times:
                    raise
    return attempt


def fetch(url: str, timeout: int) -> str:
    return f"GET {url} within {timeout}s"


patient_fetch = retry(3, fetch)
result: str = patient_fetch("https://example.test", 10)

File: tests/typesafety/typescript/higher_order.ts
// expect: passes

class ConnectionLost extends Error {}

function retry<A extends unknown[], R>(
  times: number,
  operation: (...args: A) => R,
): (...args: A) => R {
  return (...args: A): R => {
    let failures = 0;
    for (;;) {
      try {
        return operation(...args);
      } catch (error) {
        if (!(error instanceof ConnectionLost)) throw error;
        failures += 1;
        if (failures >= times) throw error;
      }
    }
  };
}

function fetch(url: string, timeout: number): string {
  return `GET ${url} within ${timeout}s`;
}

const patientFetch = retry(3, fetch);
export const result: string = patientFetch("https://example.test", 10);
